From a86629b48d974182c9cb15eacfd6d7ca7a482d94 Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Fri, 14 Sep 2018 21:20:27 +0800
Subject: [PATCH 4718/5242] MLK-19664-3 ARM64: dts: imx8qm domu: passthrough
 pcie to DomU

commit  c1fcb3fbcac15a7132ac148b932e75bdaa45d5d6 from
https://source.codeaurora.org/external/imx/linux-imx.git

Passthrough pcie to DomU to support pcie wifi.
Add the needed GPIOs in dom0 dts.
Add SMMU hook for pciea.
Modify the reserved-memory for pcie usage.
passthrough pcie related LPCG to make sure domu not panic when access
LPCG.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/freescale/fsl-imx8qm-mek-dom0.dts     |   43 +++++++++++++++++-
 .../boot/dts/freescale/fsl-imx8qm-mek-domu.dts     |   46 +++++++++++++++++---
 arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi  |   16 +++++++
 3 files changed, 99 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
index 06c9acb..d9ac30c 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0.dts
@@ -105,6 +105,11 @@
 				SC_R_DMA_0_CH14
 				SC_R_DMA_0_CH15
 				SC_R_MU_2A
+				/* pcie */
+				SC_R_PCIE_B
+				SC_R_PCIE_A
+				SC_R_SERDES_0
+				SC_R_HSIO_GPIO
 				>;
 			pads = <
 				/* i2c1_lvds1 */
@@ -131,7 +136,14 @@
 				SC_P_UART1_CTS_B
 				SC_P_UART1_RTS_B
 				SC_P_QSPI1A_DQS
+				/* pciea */
+				SC_P_PCIE_CTRL0_CLKREQ_B
+				SC_P_PCIE_CTRL0_WAKE_B
+				SC_P_PCIE_CTRL0_PERST_B
+				SC_P_LVDS1_I2C0_SDA
+				SC_P_USDHC2_RESET_B
 				>;
+			gpios = <&gpio1 13 GPIO_ACTIVE_LOW>, <&gpio4 9 GPIO_ACTIVE_LOW>, <&gpio4 29 GPIO_ACTIVE_LOW>, <&gpio4 27 GPIO_ACTIVE_LOW>;
 		};
 	};
 
@@ -181,6 +193,11 @@
 		reg = <0x0 0x5d2a0000 0x0 0x10000>;
 		xen,passthrough;
 	};
+
+	rpmsg_reserved_mem: rpmsg_reserved_mem@90000000 {
+		reg = <0x0 0x90000000 0x0 0x400000>;
+		xen,passthrough;
+	};
 };
 
 &mu_rpmsg1 {
@@ -188,6 +205,8 @@
 };
 
 &rpmsg1 {
+	/* Let xen not create mapping form dom0 */
+	/delete-property/ reg;
 	status = "disabled";
 };
 
@@ -251,7 +270,7 @@
 &smmu {
 	mmu-masters = <&dpu2 0x13>, <&gpu_3d1 0x15>,
 		      <&usdhc1 0x12>, <&usbotg1 0x11>,
-		      <&edma01 0x10>, <&cm41 0x09>;
+		      <&edma01 0x10>, <&cm41 0x09>, <&pciea 0x08>;
 };
 
 &lvds_region2 {
@@ -399,7 +418,29 @@
 	xen,passthrough;
 };
 
+&hsio {
+	xen,passthrough;
+};
+
+&hsio_pcie_x2_lpcg {
+	xen,passthrough;
+};
+
+&hsio_phy_x2_lpcg {
+	xen,passthrough;
+};
+
+&hsio_pcie_x2_crr2_lpcg {
+	xen,passthrough;
+};
+
+&hsio_pcie_x1_lpcg {
+	xen,passthrough;
+};
+
 &pciea {
+       #stream-id-cells = <1>;
+       iommus = <&smmu>;
        xen,passthrough;
 };
 
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
index a3b47a9..e6b0bf8 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu.dts
@@ -88,9 +88,13 @@
 		 *
 		 */
 
-		rpmsg_reserved: rpmsg@0x90100000 {
+		/*
+		 * CM40 rpmsg memory is still for Dom0, the domu.cfg
+		 * not map 0x90000000 - 0x90100000 to DomU.
+		 */
+		rpmsg_reserved: rpmsg@0x90000000 {
 			no-map;
-			reg = <0 0x90100000 0 0x20000>;
+			reg = <0 0x90000000 0 0x400000>;
 		};
 
 		/* global autoconfigured region for contiguous allocations */
@@ -172,6 +176,21 @@
 			compatible = "fsl,imx8qm-iomuxc";
 		};
 
+		regulators {
+			compatible = "simple-bus";
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			epdev_on: fixedregulator@100 {
+				compatible = "regulator-fixed";
+				regulator-min-microvolt = <3300000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-name = "epdev_on";
+				gpio = <&gpio4 9 0>;
+				enable-active-high;
+			};
+		};
+
 		#include "fsl-imx8qm-device.dtsi"
 
 		mu2: mu@5d1d0000 {
@@ -433,7 +452,6 @@
 /delete-node/ &edma0;
 /delete-node/ &edma2;
 /delete-node/ &edma3;
-
 &gpio0 {
 	/delete-property/ power-domains;
 	status = "disabled";
@@ -555,9 +573,17 @@
 	only-dma-mask32 = <1>;
 };
 
-/delete-node/ &hsio;
 /delete-node/ &ocotp;
-/delete-node/ &pciea;
+&pciea {
+	ext_osc = <1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pciea>;
+	disable-gpio = <&gpio1 13 GPIO_ACTIVE_LOW>;
+	reset-gpio = <&gpio4 29 GPIO_ACTIVE_LOW>;
+	clkreq-gpio = <&gpio4 27 GPIO_ACTIVE_LOW>;
+	epdev_on-supply = <&epdev_on>;
+	status = "okay";
+};
 /delete-node/ &pcieb;
 /delete-node/ &sata;
 
@@ -658,6 +684,16 @@
 				SC_P_QSPI1A_DQS_LSIO_GPIO4_IO22		0x00000021
 			>;
 		};
+
+		pinctrl_pciea: pcieagrp{
+			fsl,pins = <
+				SC_P_PCIE_CTRL0_CLKREQ_B_LSIO_GPIO4_IO27	0x04000021
+				SC_P_PCIE_CTRL0_WAKE_B_LSIO_GPIO4_IO28		0x04000021
+				SC_P_PCIE_CTRL0_PERST_B_LSIO_GPIO4_IO29		0x04000021
+				SC_P_LVDS1_I2C0_SDA_LSIO_GPIO1_IO13		0x06000000
+				SC_P_USDHC2_RESET_B_LSIO_GPIO4_IO09		0x04000021
+			>;
+		};
 	};
 };
 
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
index bd7bd52..6840c24 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
@@ -144,6 +144,22 @@
 		compatible = "fsl,imx8qm-mu-lpcg";
 		reg = <0x0 0x5d6b0000 0x0 0x10000>;
 	};
+
+	hsio_pcie_x2_lpcg: hsio_pcie_x2_lpcg@5f050000 {
+		reg = <0x0 0x5f050000 0x0 0x10000>;
+	};
+
+	hsio_pcie_x1_lpcg: hsio_pcie_x1_lpcg@5f060000 {
+		reg = <0x0 0x5f060000 0x0 0x10000>;
+	};
+
+	hsio_phy_x2_lpcg: hsio_phy_x2_lpcg@5f080000 {
+		reg = <0x0 0x5f080000 0x0 0x10000>;
+	};
+
+	hsio_pcie_x2_crr2_lpcg: hsio_phy_x2_lpcg@5f0c0000 {
+		reg = <0x0 0x5f0c0000 0x0 0x10000>;
+	};
 };
 
 /delete-node/ &edma0;
-- 
1.7.9.5

